### PR TITLE
docs: messaging architecture

### PR DESCRIPTION
**Pull request by hairyhum**
Adding architecture docs to describe messaging properties and building blocks

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://www.ockam.io/learn/how-to-guides/contributing/CONTRIBUTING#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://www.ockam.io/learn/how-to-guides/high-performance-team/conduct/).
- [x] I have accepted the Ockam [Contributor Licence Agreement](https://www.ockam.io/learn/how-to-guides/contributing/cla/) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/ockam-network/contributors/blob/master/CONTRIBUTORS.csv) file in a separate pull request to the [ockam-network/contributors](https://github.com/ockam-network/contributors) repository.

### CODE REVIEW

File: documentation/architecture/messaging/Reliability.md [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+1. Unreliable. When messages may be lost or duplicated `DM = {R < 1, UR < 1}`
+1. At most once. When all messages are unique, but may be missing `DM = {R < 1, UR = 1}`
+1. At least once. When all messages are received, but may have duplicates `DM = {R = 1, UR < 1}`
+1. Exactly once. When all messages are received and only once `DM = {R = 1, UR = 1}`
+
+Exactly once delivery is controversal and hard to achieve in presence of errors.
+For practical purposes we're going to use more relaxed version of this definition
+where `DM = {R -> 1, UR -> 1}`
+
+<img src="./images/delivery_modes.jpg" width="100%">
```
Comment (dvermd) [inline documentation/architecture/messaging/Reliability.md:42]: There's a repeated typo in this image on `receivzed`

File: documentation/architecture/messaging/Reliability.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+then we can describe the following modes:
+
+1. Unreliable. When messages may be lost or duplicated `DM = {R < 1, UR < 1}`
+1. At most once. When all messages are unique, but may be missing `DM = {R < 1, UR = 1}`
+1. At least once. When all messages are received, but may have duplicates `DM = {R = 1, UR < 1}`
+1. Exactly once. When all messages are received and only once `DM = {R = 1, UR = 1}`
+
+Exactly once delivery is controversal and hard to achieve in presence of errors.
+For practical purposes we're going to use more relaxed version of this definition
+where `DM = {R -> 1, UR -> 1}`
```
Comment (dvermd) [inline documentation/architecture/messaging/Reliability.md:40]: Could you clarify what the notation '->' means in `DM = {R -> 1, UR -> 1}` ?

File: documentation/architecture/messaging/Reliability.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+Retries decrease uniqness rate
+Given `DM(A->B) = {0.5, 1}`
+`DMr(A->B)` - delivery mode of `A->B` with retries
+`DMr∞(A->B) = {->1, ->0.5}`
+
+To control retries, confirmation messages and timeouts are used
+
+<img src="./images/retries.jpg" width="100%">
+
+To provide at-least once delivery between workers which don't have additional logic,
```
Comment (dvermd) [inline documentation/architecture/messaging/Reliability.md:108]: You may quote or use any other artifact to visually group the delivery types
```suggestion
To provide `at-least once` delivery between workers which don't have additional logic,
```

File: documentation/architecture/messaging/Reliability.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+<img src="./images/pipe_injection.jpg" width="100%">
+
+### Cascade confirmation
+
+In case we don't have a reliable delivery to the last step or we can have errors processing the message,
+then the pipe receiver may be configured to send confirms to the pipe sender only after it gets a confirm from the next worker in the pipeline
+
+For example if we have a pipeline `A->P1; P1->P2; P2->B` and delivery `P2->B` is not reliable, then `P2` should only send confirms to `P1` when it receives a confirm from `B`
+
+There can be multiple confirming workers in this pipeline, e.g. `A->P1; P1->P2; P2->P3; P3-B` etc
```
Comment (dvermd) [inline documentation/architecture/messaging/Reliability.md:164]:
```suggestion
There can be multiple confirming workers in this pipeline, e.g. `A->P1; P1->P2; P2->P3; P3->B` etc
```

File: documentation/architecture/messaging/Delivery.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+more on pipes in [Pipes and Channels](./Pipes_Channels.md)
+
+
+## Properties
+
+1. [Accessibility](./Accessibility.md)
+  Accessibility describes an ability of the route (and workers in this route) to deliver messages to the destination.
+  Accessibility is required property in messaging, since there is not delivery without it.
+
+1. [Reliability and uniqness](./Reliability.md)
```
Comment (dvermd) [inline documentation/architecture/messaging/Delivery.md:56]:
```suggestion
1. [Reliability and uniqueness](./Reliability.md)
```
you may `sed` that one

File: documentation/architecture/messaging/Pipes_Channels.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+### Accessibility and session setup
+
+Pipes:
+
+In order to have coordinated state, sender and receiver must be accessible by each other. At least receiver should be accessible by the sender.
+
+Since messages are flowing from the sender, receiver does not initiate the communication and rarely needs to store a route to the sender in its state.
+
+1. Static pipe - a pipe in which sender is created with a known route to the receiver
+2. Session pipe - a pipe in which sender is created with a session init route and receiver mey be created during the session establishment
```
Comment (dvermd) [inline documentation/architecture/messaging/Pipes_Channels.md:80]:
```suggestion
2. Session pipe - a pipe in which sender is created with a session init route and receiver may be created during the session establishment
```

File: documentation/architecture/messaging/Pipes_Channels.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    RR: return_route
+  ```
+
+Pipes facilitate unidirectional message delivery.
+Pipes preserve the return route of the original message BUT do not trace a return route
+
+Pipes may use a coordinated state and send more messages between sender and receiver to provide
+additional message delivery properties.
+
+<img src="./Pipe.jpg" width="100%">
```
Comment (dvermd) [inline documentation/architecture/messaging/Pipes_Channels.md:34]: Images are missing

**Review by dvermd — COMMENTED** [review]
This a great piece of architecture

File: documentation/architecture/messaging/Accessibility.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+<img src="./images/pipe.jpg" width="100%">
+
+This combination of workers is called a **routing pipe**
+
+From the perspective of `A`, routing pipe works the same as a single route-based forwarding worker,
+which allows an abstraction to hide `Fst1->Fr2` communication
+
+Pipes are used a lot in Ockam messaging, more on pipes in [Pipes and Channels](./Pipes_Channels.md)
+
+## Mutual accessibility with return routes
```
[the remark this answers is not in the document]
Comment (hairyhum) [reply documentation/architecture/messaging/Accessibility.md:110]: Yes. Backtracing is enough to make workers mutually accessible, but it's not required.
Even more, with some state on both ends you can configure forwarding workers to create backtracing routes even if routes in the middle are not backtracing.
For example our Stream channel implementations: messages in the stream are not tracing return route, but add the "reply stream" metadata to the messages.

File: documentation/architecture/messaging/Reliability.md [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+Duplication rate of `1` means each received message was received only once, rate of `2` means that messages were received 2 times on average, not necessary each message was duplicated though.
+
+For consistency wirh reliability we can use inverted duplication rate, call it **uniqness rate** as `UR = 1/DR = Ru/R`
+
+If `UR = 1` when each message was received only one time, `UR = 0.5` when messages are received twice on average
+
+<img src="./images/reliability.jpg" width="100%">
+
+Reliability `R(A->B) = 3/4` - 1 message out of 4 was lost
+Uniqness rate `UR(A->B) = 3/4` - 1 message out of 4 was duplicated
```
[the remark this answers is not in the document]
Comment (hairyhum) [reply documentation/architecture/messaging/Reliability.md:23]: Not sure I follow. This is a statistical value, so if you lose every 1 of 2 messages, you likely will lose 2 of 4.

File: documentation/architecture/messaging/Reliability.md (hunk 1 above)
Comment (hairyhum) [reply documentation/architecture/messaging/Reliability.md:44]: Can't find it. Must have been addressed already.

File: documentation/architecture/messaging/Reliability.md (hunk 3 above)
Comment (hairyhum) [reply documentation/architecture/messaging/Reliability.md]: For consistency will use at-least-once, at-most-once and exactly-once without quoting.

File: documentation/architecture/messaging/Reliability.md (hunk 2 above)
Comment (hairyhum) [reply documentation/architecture/messaging/Reliability.md]: Added a bit of explanation there.

**Comment by hairyhum** [comment]
@dvermd Thanks for review.

**Review by mrinalwadhwa — APPROVED** [review]
Looks great, let's merge these 👍

**Pull request merged**